### DISCUSSION TITLE
Docker task type

### DISCUSSION THREAD

**Discussion by simonwo** [ucan-wg/invocation#5] [Ideas]
This is the task type that Bacalhau will be using to describe Docker jobs. Is this similar enough to anyone else that it is worth joining up on it? How closely does it match what other people are using, and how could we tweak it so it can work for you?

## Docker invocations
A Docker invocation is identified by a UCAN Ability of `docker/run`. The resource specified SHOULD then be interpreted as a Docker image identifier. 

### Image references
The image identifier SHOULD be a fully-resolved image reference (e.g. `ubuntu@sha256:4b1d0c4a2d2aaf63b37111f34eb9fa89fa1bf53dd6e4ca954d47caebca4005c2`). 

* References MAY include the repository reference (e.g. `docker.io/ubuntu@sha256:4b1d0c4a2d2aaf63b37111f34eb9fa89fa1bf53dd6e4ca954d47caebca4005c2`) 
* References MAY use URIs from other schemes in place of image references (e.g. `ipfs://Qmyd0ck3rc0ntainer`). 
* Implementations MAY support label tags (e.g. `ubuntu@latest`).

### Inputs schema
For a Docker invocation, the `inputs` structure SHOULD conform to the following schema:
```ipldsch
type DockerInputs struct {
    entrypoint [String]
    workdir String
    mounts {String:Any}
    outputs {String:Any}
    env {String:String}
    workdir String
}
```

where the fields are defined as:

| Name       | Type            | Description                                                       |
|------------|-----------------|-------------------------------------------------------------------|
| entrypoint | [String]        | The program and arguments that will be invoked in the container   |
| workdir    | String          | Path within the container in which the entrypoint will be invoked |
| mounts     | {String:Any} | Read-only disk mounts that will contain the data to be stored. The value of the map should be an object describing the data to be mounted, e.g. a CID or URL, or null for an empty mount.     |
| outputs    | {String:Any} | Writeable disk mounts, initially empty. The value of the map should either null for an empty mount or executor-specific options for how to handle the outputs.                            |
| env        | {String:String} | Environment variables made available to the container             |
| workdir    | String          | Path within the container in which the entrypoint will be invoked |

### Output schema
For a Docker invocation, a successful result should match the schema:
```ipldsch
type DockerOutput struct {
    stdout String
    stderr String
    exitcode Integer
    outputs {String:Any}
}
```

where the fields are defined as:

| Name       | Type            | Description                                                       |
|------------|-----------------|-------------------------------------------------------------------|
| stdout | String        | Text written to the standard out stream   |
| stderr    | String          | Text written to the standard error stream |
| exitcode     | Integer | The exit code returned from invoking the entrypoint within the container     |
| outputs    | {String:Any} | The final contents of the disk mounts specified in the inputs. The value of the map should be an object describing the output data, e.g. a CID or URI.                          |

**Comment by expede** [comment]
@simonwo this looks great 👍 💯 I'll give it some deeper thought, but if the above works for you, then this fits really well into the model 🚀
